core/src/main/java/org/ehcache/util/ConcurrentWeakIdentityHashMap.java

```diff
@@ -117,15 +119,56 @@ public void clear() {
 
   @Override
   public Set<K> keySet() {
-    purgeKeys();
-    final HashSet<K> ks = new HashSet<K>();
-    for (WeakReference<K> kWeakReference : map.keySet()) {
-      final K k = kWeakReference.get();
-      if (k != null) {
-        ks.add(k);
+    return new AbstractSet<K>() {
+      @Override
+      public Iterator<K> iterator() {
+        purgeKeys();
+        final Iterator<WeakReference<K>> iterator = map.keySet().iterator();
+        final AtomicReference<K> ref = new AtomicReference<K>();
```

Comment:
Because of the required final to use it inside the anonymous iterator. Or do you suggest that a simple array would be better?